### PR TITLE
Open Generic services don't work with anything but implementation types

### LINKED ISSUE #257: Open Generic services don't work with anything but implementation types
The generic service registration assumes an implementation type and explodes with a null ref if one wasn't provided. We should either throw up front or support this in some way. Supporting an implementation factory would mean that the callback would need access to the closed generic.

/cc @lodejard

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #257

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @davidfowl \ @Eilon

File: src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceDescriptor.cs [hunk 1]
```diff
@@ -22,6 +23,14 @@ public class ServiceDescriptor
                                  ServiceLifetime lifetime)
             : this(serviceType, lifetime)
         {
+            if (serviceType.GetTypeInfo().IsGenericTypeDefinition &&
```
Comment (pranavkm) [inline src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceDescriptor.cs]: Thoughts on verifying it in the `ServiceDescriptor`? Seemed like the one place to all the registrations get funneled to.

Comment (Eilon) [inline src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceDescriptor.cs]: I don't really know any better, but that sounds fine to me.

**Comment by Eilon** [comment]
:shipit:

**Comment by davidfowl** [comment]
Not sure this fix is in the right place. What if this is supported on other DI implementations? The thing that throws the exception today is the default implementation. The descriptor is just metadata. This is where it should throw from:

https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceTable.cs#L27

**Comment by pranavkm** [comment]
Fair enough.

**Comment by pranavkm** [comment]
Updated

**Comment by Eilon** [comment]
Ok still looks good to me... :shipit:

**Pull request closed without merging**